File: backend/app/models/playlist.py
"""Playlist model for storing managed playlist configurations."""

from datetime import datetime
from enum import Enum as PyEnum

from sqlalchemy import Boolean, ForeignKey, Integer, String, UniqueConstraint
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.sql import func

from app.database import Base
from app.models.types import UTCDateTime

# ``episode_limit`` encoding, shared by the playlist default and the
# per-assignment override: 0 means "all unplayed", n >= 1 means "at most n".
# Using the same encoding at both levels leaves NULL free to mean "inherit"
# on the assignment (docs/design/assignment-rules.md).
ALL_EPISODES = 0


class PickFrom(str, PyEnum):
    """Which end of a show's unplayed episodes to take from.

    Also the order the show's episodes are listened to. ``OLDEST`` is what a
    serial wants ("next unfinished"); ``NEWEST`` is what daily news wants.
    """

    NEWEST = "newest"
    OLDEST = "oldest"


class Arrangement(str, PyEnum):
    """How the contributions of each assigned show are assembled."""

    BY_POSITION = "by_position"  # groups in assignment order
    BY_DATE = "by_date"  # everything merged by release date
    SHUFFLE = "shuffle"  # shows interleaved at random, each show kept in order


class DateDirection(str, PyEnum):
    """Direction of the merge when ``arrangement`` is ``BY_DATE``."""

    NEWEST_FIRST = "newest_first"
    OLDEST_FIRST = "oldest_first"


class Playlist(Base):
    """Playlist model storing configuration for managed playlists.

    The playlist owns the *defaults* for its assignments and the *assembly*
    rules. What each show contributes is decided per assignment — see
    ``PlaylistPodcast`` and ``services/assignment_rules.py``.
    """

    __tablename__ = "playlists"
    # Two playlists linked to one Spotify playlist would overwrite each other
    # on every rebuild. The API refuses it up front; this is what makes the
    # refusal race-proof (issue #245). NULL repeats freely — SQLite and
    # Postgres both treat NULLs as distinct — so unlinked playlists are fine.
    __table_args__ = (UniqueConstraint("user_id", "spotify_playlist_id", name="uq_playlists_user_spotify_playlist"),)

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    user_id: Mapped[int] = mapped_column(Integer, ForeignKey("users.id", ondelete="CASCADE"), nullable=False)
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    spotify_playlist_id: Mapped[str | None] = mapped_column(String(255), nullable=True)
    is_enabled: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)

    # Defaults inherited by assignments that have no override.
    default_episode_limit: Mapped[int] = mapped_column(Integer, default=ALL_EPISODES, nullable=False)
    default_pick_from: Mapped[str] = mapped_column(String(10), default=PickFrom.NEWEST.value, nullable=False)

    # Assembly.
    arrangement: Mapped[str] = mapped_column(String(20), default=Arrangement.BY_POSITION.value, nullable=False)
    date_direction: Mapped[str] = mapped_column(String(20), default=DateDirection.OLDEST_FIRST.value, nullable=False)

    # Tracking
    last_updated_at: Mapped[datetime | None] = mapped_column(UTCDateTime, nullable=True)

    # Timestamps
    created_at: Mapped[datetime] = mapped_column(UTCDateTime, server_default=func.now(), nullable=False)

    # Relationships
    podcast_assignments = relationship("PlaylistPodcast", back_populates="playlist", cascade="all, delete-orphan")

    def __repr__(self) -> str:
        return f"<Playlist(id={self.id}, name={self.name}, arrangement={self.arrangement})>"

File: ios/PodcastManager/PodcastManager/Models/Playlist.swift
import Foundation

struct Playlist: Codable, Identifiable, Hashable {
    let id: Int
    let name: String
    let isEnabled: Bool
    /// Default episodes per podcast: 0 = all unplayed, n >= 1 = at most n.
    let defaultEpisodeLimit: Int
    /// Which end of a show's unplayed episodes to take from: "newest" | "oldest".
    let defaultPickFrom: String
    /// How contributions are assembled: "by_position" | "by_date" | "shuffle".
    let arrangement: String
    /// Only used when arrangement is "by_date": "newest_first" | "oldest_first".
    let dateDirection: String
    let spotifyPlaylistId: String?
    let lastUpdatedAt: String?
    let createdAt: String
    let podcastCount: Int

    /// e.g. "All unplayed", "Latest only · newest", "Up to 3 · oldest".
    var ruleSummary: String {
        episodeRuleSummary(limit: defaultEpisodeLimit, pickFrom: defaultPickFrom)
    }

    /// Public Spotify URL, or nil until the first run has created the playlist.
    var spotifyURL: URL? {
        guard let id = spotifyPlaylistId, !id.isEmpty else { return nil }
        return URL(string: "https://open.spotify.com/playlist/\(id)")
    }

    var arrangementLabel: String {
        switch arrangement {
        case "by_date":
            return dateDirection == "oldest_first" ? "By date, oldest first" : "By date, newest first"
        case "shuffle":
            return "Shuffled"
        default:
            return "Podcast order"
        }
    }
}

/// Shared phrasing for playlist defaults and resolved assignment rules.
func episodeRuleSummary(limit: Int, pickFrom: String) -> String {
    let direction = pickFrom == "oldest" ? "oldest" : "newest"
    // The direction is shown even when unlimited: it still sets the order the
    // show's episodes are listened to within its group (matches the web UI).
    switch limit {
    case 0: return "All unplayed · \(direction)"
    case 1: return "Latest only · \(direction)"
    default: return "Up to \(limit) · \(direction)"
    }
}

struct PlaylistListResponse: Codable {
    let items: [Playlist]
    let total: Int
}

struct PlaylistRunResponse: Codable {
    let message: String
    let playlistId: Int
    let episodeCount: Int
    /// Disabled playlist left untouched.
    let skipped: Bool?
    /// Playlist was written, but from incomplete data.
    let partial: Bool?
}

struct MessageResponse: Codable {
    let message: String
}

struct PlaylistRunAllResponse: Codable {
    let message: String
    let results: [PlaylistRunResult]
}

struct PlaylistRunResult: Codable, Identifiable {
    let playlistId: Int
    let playlistName: String
    let success: Bool
    let episodeCount: Int?
    let error: String?
    let skipped: Bool?
    let partial: Bool?

    var id: Int { playlistId }
}
